### PR TITLE
Tests for WaitingForResources and Executing

### PR DESCRIPTION
**Pull request by rmetzger**
Looking for some quick feedback (@tillrohrmann @zentol) on the tests, before I start adding them for all states.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        public void setHasEnoughResources(Supplier<Boolean> sup) {
+            hasEnoughResourcesSupplier = sup;
+        }
+
+        public void setCreateExecutionGraphWithAvailableResources(Supplier<ExecutionGraph> sup) {
+            this.createExecutionGraphWithAvailableResources = sup;
+        }
+    }
+
+    private static class MockArchivedExecutionGraph extends ArchivedExecutionGraph {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: there's an ArchivedExecutionGraphBuilder

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    /** WaitingForResources is transitioning to Executing if resources are stable. */
+    @Test
+    public void testTransitionToExecuting() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> true);
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.EXECUTING, JobStatus.RUNNING)));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: This seems to be mostly testing the MockContext implementation. Shouldn't we just test that it is calling `toExecuting`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+/** Tests for the WaitingForResources state. */
+public class WaitingForResourcesTest extends TestLogger {
+
+    /** WaitingForResources is transitioning to Executing if resources are stable. */
+    @Test
+    public void testTransitionToExecuting() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> true);
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: side-note: en empty amount of required resources should fail with an IllegalArgumentException.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the WaitingForResources state. */
+public class WaitingForResourcesTest extends TestLogger {
+
+    /** WaitingForResources is transitioning to Executing if resources are stable. */
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: We aren't testing whether resources are stable; just that they are sufficient.

**Review by zentol — CHANGES_REQUESTED** [review]
I don't think the tests really highlight what is being tested; it's all boils down to some assertion on this list of transitions, where we check some enum values that are mostly set by the context.
It's not really intuitive to me that in order to verify that `goToExecuting` was called I have to assert that the values set by the MockContext showed up.
Something that could maybe work well is changing the MockContext to accept some form of expectation for the different state transitions, like a consumer (presumably doing some matching on the arguments), that fails if the method was never called.

We don't need a list of state transitions; there should only ever be a single transition; if more than one is triggered we can fail right away.

The explicit calls to `onEnter` are likely to introduce a defect at some point, and we seem to be ignoring the `onLeave` hook. I'm not really sure what the solution is, but ideally we figure out a way to do this such that these are always called if necessary (e.g., if `goToExecuting` is called).
Maybe some form of `AutoCloseable` `TestingStateMachine` which accepts a state in it's constructor (calling `onEnter`) and triggering `onLeave` if any state transition was run (checked via `getStateTransitions` or maybe notifications from the context?

**Comment by rmetzger** [comment]
Thanks a lot for your feedback!

I've tried coming up with a better approach. I completely overlooked the `.onLeave()` call (which has quite a lot of dependencies) in the first try.
This second try is not ideal yet, but should handle the onEnter / onLeave by the testing framework.
Also, there's no more list of state transitions, and the tests are really reduced to a minimum.

More feedback is always welcome .. I'll try to continue working on it tonight or tomorrow morning to maybe come up with improvements myself.
I probably need to move some of the inner classes to new files, and share them between the tests for the different states (at least the mocked contexts).

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+            waitingForResources.onLeave(target);
+        }
+
+        private State getTransitionTo() {
+            return transitionTo;
+        }
+
+        @Override
+        public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {
+            transitionToState(new Finished(null, archivedExecutionGraph, LOG));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: I'd say that this goes too far. These are scheduler implementation details (e.g., how the `Finished` state is represented), or what happens within `goToFinished` in it's entirety.

This path will likely lead us toward just another state machine.

Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: I was thinking more of something like this:

```
class MockContext ... implements AutoCloseable

void setExpectFinished(Consumer<EG> asserter) {
    this.noFinishTrap=() -> throw new AssertionError("no transition to finished");
    this.setGoToExecutingConsumer(asserter);
}

void goToExecuting(EG eg) {
   this.noFinishTrap = ()->{};
   this.didTransition=true;
   // assert arguments
   this.goToExecutinConsumer.accept(eg);
}

void close() {
   // noop if expected transition occurred
   this.noFinishTrap.run();
}}

class TestStateContext implements AutoCloseable {
  final MockContext context
  final WaitingForResources wfr

void close()
  if this.context.didTransition();
    this.wfr.onLeave();
  this.context.close()
}

class *Test

  @Test
  void testSomething()
    ctx = ...
    wfr =  ...

    try (tsc = new TSC(ctx, wfr)) {
      ctx.expectFinished(...)
      wfr.<doSOmethingThatResultsInFinished>
    }
```

Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: > I'd say that this goes too far. These are scheduler implementation details (e.g., how the Finished state is represented), or what happens within goToFinished in it's entirety.

I would have really liked to avoid this, but I added it because we need a `Finished` instance to pass to the `onLeave(State)` method.

Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: Also thanks for the code mockup. What I like about it is that the tested class is instantiated in the test itself. My approach hides everything, and reduces the test only to the required input (making the tests shorter, but also more abstract).

I'll try to implement a few tests with your approach to see how well it works

Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: Oh, I didn't even notice that onLeave gets the _new state_ ... ideally we get rid of that. Maybe we can tie to disposal of the operator coordinator to the EG termination future instead.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        wfr.suspend(new Exception("test"));
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.FINISHED, JobStatus.SUSPENDED)));
+    } */
+
+    private class StateTestContext {
+        private final MockContext mockContext;
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: For my taste, two contexts are one too many. Maybe give it another name.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        private WaitingForResources waitingForResources;
+
+        public StateTestContext(Consumer<MockContext> modifyContext) {
+            this.mockContext = new MockContext();
+            modifyContext.accept(mockContext);
+            this.waitingForResources =
+                    new WaitingForResources(mockContext, log, ResourceCounter.empty());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: I don't like that the subject under test is instantiated in some other helper class. When looking at the test I want to see how stuff is being set up.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+
+    private static class MockContext implements WaitingForResources.Context {
+
+        private static final Logger LOG = LoggerFactory.getLogger(MockContext.class);
+
+        private WaitingForResources waitingForResources;
+
+        private State transitionTo = null;
+        private Supplier<Boolean> hasEnoughResourcesSupplier = () -> false;
+        private Supplier<ExecutionGraph> createExecutionGraphWithAvailableResources =
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]:
```suggestion
        private SupplierWithException<ExecutionGraph> createExecutionGraphWithAvailableResources =
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+            @Override
+            public Executor getMainThreadExecutor() {
+                return ForkJoinPool.commonPool();
+            }
+
+            @Override
+            public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {}
+        }
+    }
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: This seems wrong. We shouldn't have to implement something for the `Executing` state when testing the `WaitingForResources` test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+
+    private static class MockContext implements WaitingForResources.Context {
+
+        private static final Logger LOG = LoggerFactory.getLogger(MockContext.class);
+
+        private WaitingForResources waitingForResources;
+
+        private State transitionTo = null;
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: I would suggest to not store the last state here. This effectively couples all states together.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+        // no resources --> no state transition
+        testContext.getWaitingForResources().notifyNewResourcesAvailable();
+        testContext.assertNoStateTransition();
+
+        // make resources available
+        testContext.getMockContext().setHasEnoughResources(() -> true);
+        testContext.getWaitingForResources().notifyNewResourcesAvailable();
+
+        testContext.assertStateTransitionTo(Executing.class);
+    }
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: This test can be split into the two cases

1. not enough resources
2. enough resources

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: `onLeave(State)` is indeed not optimal. This somewhat couples those implementations which rely on a specific State implementation in the `onLeave` method. Maybe we can test this method in isolation and with more of an integration test.

The problem `onLeave(State)` should solve is that we need to close some services when leaving a certain state. One way to solve it could be to pass this service to the context method and let the implementor of this context decide how to handle the service. The downside is that now every state transition would get the full set of services.

Alternatively, the state might need to remember into which state it transitioned and depending on this close or leave the service when `onLeave()` is called (w/o the new state). But this seems brittle.

I think that Chesnay's proposal should work.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+            actions.add(Tuple3.of(expectedState, action, delay));
+        }
+
+        // ---- Testing extensions ------
+
+        public void setWaitingForResources(WaitingForResources wfr) {
+            this.waitingForResources = wfr;
+        }
+
+        public List<Tuple3<State, Runnable, Duration>> getActions() {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java]: I guess you are using tuples because it is a wip branch. In the final PR, please don't use tuples. They make reading code unnecessarily hard.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @rmetzger. I left a couple of comments.

**Comment by rmetzger** [comment]
Thank you both for the great feedback!
I pushed an updated version, now also including tests for `Executing`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+                                        .getExceptionAsString()
+                                        .contains(testExceptionString));
+                    });
+
+            wfr.handleGlobalFailure(new RuntimeException(testExceptionString));
+        }
+    }
+
+    @Test
+    public void testCancel() throws Exception {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java:148]: hmm this test is a bit weird. After we transition to Executing then no further operations should be executed on the WaitingForResources state.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+            wfr.cancel();
+        }
+    }
+
+    @Test
+    public void testSuspend() throws Exception {
+        try (MockContext ctx = new MockContext()) {
+            ctx.setHasEnoughResources(() -> true);
+            ctx.setExpectExecuting(assertNonNull());
+            ctx.setExpectFinished(
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java:166]: same as above

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+        public ExecutionGraphHandler getExecutionGraphHandler() {
+            return executionGraphHandler;
+        }
+
+        public OperatorCoordinatorHandler getOperatorCoordinatorHandle() {
+            return operatorCoordinatorHandle;
+        }
+    }
+
+    private static class RestartingArguments {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: nit: this could extend `CancellingArguments` to deduplicate things a bit.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
+            ctx.setGetMainThreadExecutor(() -> executor);
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+            exec.onEnter();
+            // transition EG into terminal state, which will notify the Executing state about the
+            // failure (async via the supplied executor)
+            exec.getExecutionGraph().failJob(new RuntimeException("test failure"));
+            executor.shutdown();
+            executor.awaitTermination(10, TimeUnit.SECONDS);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: this should probably be moved into a finally block

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnSuspend() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFinished(
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: hmm...now that I see it in action I'm wondering...like, we _could_ just capture the state transition in the context and then do an assert on it, right? We don't actually need all this trap business that I proposed.

Structurally, the test would then look like this, which is more in line with our usual test patterns:
```
try (MockExecutingContext ctx = new MockExecutingContext()) {
    Executing exec = getExecutingState(ctx);
    exec.onEnter();
    exec.suspend(new RuntimeException("suspend"));

    ctx.assertFinished(
            archivedExecutionGraph -> {
                assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
            });
}
```
where the assert fails immediately if the transition did not occur (i.e., the *Arguments are null).

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+        @Override
+        public void runIfState(State expectedState, Runnable action) {
+            if (expectedStateChecker.apply(expectedState)) {
+                action.run();
+            }
+        }
+
+        @Override
+        public boolean isState(State expectedState) {
+            return false;
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: This should return `true` until we transitioned to another state

**Review by zentol — COMMENTED** [review]
Some preliminary feedback, still need to dive into the test specifics.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 16 above)
Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: The problem is that the AutoClosable.close is called before executing the finally block, hence the test fails because the state transition hasn't happened yet.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 17 above)
Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: This is a bit more intuitive to read, yes. But I believe it leads to longer test methods (because we always need to assert something state related in the end (I don't think using some fancy logic in AutoClosable.close() to decide if no or only one asserted state transition happened makes sense))

The current (PR) approach has the benefit that it fails if an unexpected state transition is happening. We would need to introduce the methods `ctx.assertOnlyFinished(..)` and  `ctx.assertNoStateTransition()`.

Also, I don't think the implementation of the context would *really* benefit, might actually be more tedious:
image

I'm leaning towards leaving it as-is.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 18 above)
Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: This method is actually not covered by the tests (since it's used when triggeringSavepoints, which I suppose should be tested in an declarative scheduler IT case). I will throw and exception if this method gets called.

**Comment by rmetzger** [comment]
Thanks a lot for your quick feedback. I addressed all your remarks!

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 17 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: > because we always need to assert something state related in the end

I don't think we'd need to; as-is if we don't setup an expectation then a state transition might occur without the test failing. Why would we always need an assertion now?

> [we'd need] ctx.assertOnlyFinished(..)

We should in general enforce that only 1 state transition is triggered. If this happens it is either a bug in the state implementation or in the test itself.

> [we'd need] ctx.assertNoStateTransition()

I'd think we'll need that at some point anyway.

> context [...] might actually be more tedious:

You'd only need a shared helper method checking that only 1 of the arguments is not-null. Coupled with the initial assertion this ensures that only the transition in question has occurred.
Not even that; you just need a boolean flag that is set if the state transition.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 16 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: Does that mean we are relying on the executor shutdown to run outstanding tasks? If so, then we should make this more explicit (for example with a `ManuallyTriggeredScheduledExecutorService`) because I did not expect that from a quick glance.

Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: Allright, I'll update accordingly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java (hunk 17 above)
Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java]: Okay, I migrated the ExecutingTests to the new model.

**Comment by rmetzger** [comment]
Let's continue reviewing this on the Flink repo: https://github.com/apache/flink/pull/14852/files
